### PR TITLE
feat(ui): Add Suspend and Resume for Cronworkflow in UI. Closes #1975

### LINKED ISSUE #1975: Add Suspend and Resume for  Cronworkflow in UI
Provide the Suspend and Resume functionality in Cronworkflow UI

### PR DESCRIPTION
**Pull request by changhc**
#### Screenshot
image

Updated:
image

Closes #1975.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to USERS.md.

### CODE REVIEW

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx [hunk 1]
```diff
@@ -93,4 +109,32 @@ export class CronWorkflowDetails extends BasePage<RouteComponentProps<any>, Stat
...
+        wf.spec.suspend = true;
+        services.cronWorkflows
+            .update(wf, this.name, this.namespace)
+            .catch(e => {
+                this.appContext.apis.notifications.show({
+                    content: 'Failed to suspend cron workflow ' + e,
+                    type: NotificationType.Error
+                });
+            })
+            .then(() => this.getCronWorkflow());
```
Comment (simster7) [inline ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:124]: Doesn't the `update` function return the new, updated Workflow? I don't think we need to make this extra API call

Comment (changhc) [reply ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx]: You're right. I've updated this part. Thanks!

**Comment by changhc** [comment]
I discovered some bugs while addressing the comment. They were mainly due to that the content in the editor wasn't updated when the cronworkflow was updated. I've already fixed this.

File: ui/src/app/shared/components/yaml/yaml-editor.tsx [hunk 2]
```diff
@@ -41,7 +42,7 @@ export class YamlEditor<T> extends React.Component<Props<T>, State> {
                         autoFocus={true}
                     />
                 ) : (
-                    <YamlViewer value={this.state.value} />
+                    <YamlViewer value={jsYaml.dump(this.props.value)} />
```
Comment (simster7) [inline ui/src/app/shared/components/yaml/yaml-editor.tsx:45]: Why is this change necessary?

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx [hunk 3]
```diff
@@ -8,10 +8,12 @@ import {Loading} from '../../../shared/components/loading';
 import {services} from '../../../shared/services';
 import {CronWorkflowSummaryPanel} from '../cron-workflow-summary-panel';
 
+const jsonMergePatch = require('json-merge-patch');
 require('../../../workflows/components/workflow-details/workflow-details.scss');
 
 interface State {
     cronWorkflow?: CronWorkflow;
+    editing: boolean;
```
Comment (simster7) [inline ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:16]: I'm not too sure we need this... it would make sense to me if I was editing the YAML and _without saving my edits_ pressed Suspend/Resume that my changes would be lost.

File: ui/src/app/shared/components/yaml/yaml-editor.tsx (hunk 2 above)
Comment (changhc) [reply ui/src/app/shared/components/yaml/yaml-editor.tsx:45]: Consider the following use case:
* A cronworkflow is running
* A user clicks `suspend` and the cronworkflow is suspended
* The user then edits the workflow and submits the updated version

Without this change, the update will fail since the suspended workflow is not reflected in the editor, i.e., `suspend: true` is missing, and an error will be returned asking the user to submit with a newer version of the cronworkflow.

Therefore, we need to use the prop value to display the updated value.

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx (hunk 3 above)
Comment (changhc) [reply ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:16]: This makes sense too, but I think this is safer. The ideal design would be having an additional `cancel` button besides the `submit` button while still disabling the suspend button during editing. I think this flow is more natural and ensures safety simultaneously.

File: ui/src/app/shared/components/yaml/yaml-editor.tsx (hunk 2 above)
Comment (simster7) [reply ui/src/app/shared/components/yaml/yaml-editor.tsx:45]: Correct, but if `this.props.value` is changed, the entire react component would re-render. I.e., the constructor would get called again and `this.state.value` would get filled with the new prop value. This is a feature of react.

Do you still think we need this change?

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx (hunk 3 above)
Comment (simster7) [reply ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:16]: I agree that it's safer, but I can see some potential for edge case bugs. For example, in this implementation this flag is only set to false if the underlying object is changed. Does this flag work correctly if a user clicks edit then save without making any changes?

I would rather move forward without this change for now, and add it later if necessary. Let me know if you feel strongly otherwise.

File: ui/src/app/shared/components/yaml/yaml-editor.tsx (hunk 2 above)
Comment (changhc) [reply ui/src/app/shared/components/yaml/yaml-editor.tsx:45]: Initially I had the same thought as yours, but it turned out that the editor didn't rerender.

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx (hunk 3 above)
Comment (changhc) [reply ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:16]: You're right. To solve this, we should not set `editing: false` in `onChange` but do this in another function and pass it in as a separate props. This should be called after `.then(updated => props.onChange(updated))`.

I don't feel this is troublesome or will severely affect other things but can improve UX a little bit, so to me it's still beneficial. If you still have concerns, we can skip this now.

File: ui/src/app/shared/components/yaml/yaml-editor.tsx (hunk 2 above)
Comment (simster7) [reply ui/src/app/shared/components/yaml/yaml-editor.tsx:45]: In this case can we use this solution: https://stackoverflow.com/a/51309488? I'd like us to stay consistent in using `this.state.value` across the file.

File: ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx (hunk 3 above)
Comment (simster7) [reply ui/src/app/cron-workflows/components/cron-workflow-details/cron-workflow-details.tsx:16]: Let's skip this for now please. You're welcome to open a separate PR with this change, though. If you do, please make sure that it's consistent across different resource kinds (`Workflow`, `WorkflowTemplate`, etc.)

File: ui/src/app/shared/components/yaml/yaml-editor.tsx (hunk 2 above)
Comment (changhc) [reply ui/src/app/shared/components/yaml/yaml-editor.tsx]: Updated

**Comment by simster7** [comment]
@alexec Seems like CircleCI is using a dev version of Mockery? See `codegen` check

**Comment by simster7** [comment]
@changhc Can you merge with master to fix the CI?
